calliope/preprocess/model_run.py

```diff
@@ -536,39 +599,94 @@ def _parser(x, dtformat):
 
     # Ensure all timeseries have the same index
     indices = [
-        (file, df.index)
-        for file, df in timeseries_data.items()
-        if file not in cluster_filenames
+        (tskey, df.index)
+        for tskey, df in timeseries_data.items()
+        if tskey not in cluster_filenames | cluster_dfnames
     ]
-    first_file, first_index = indices[0]
-    for file, idx in indices[1:]:
+    first_tskey, first_index = indices[0]
+    for tskey, idx in indices[1:]:
         if not first_index.equals(idx):
             raise exceptions.ModelError(
                 "Time series indices do not match "
-                "between {} and {}".format(first_file, file)
+                "between {} and {}".format(first_tskey, tskey)
             )
 
     return timeseries_data, first_index
 
 
-def _assert_timeseries_available(filenames):
-    if len(filenames) == 0:
+def _assert_either_file_or_dataframes(
+    constraint_filenames,
+    cluster_filenames,
+    constraint_dfnames,
+    cluster_dfnames,
+    timeseries_dataframes,
+):
+    if (
+        timeseries_dataframes is not None
+        and len(constraint_filenames | cluster_filenames) > 0
+    ):
+        raise exceptions.ModelError(
```

Comment:
I wonder if allowing loading either from df or file would be fine? Our process at the moment is:

find files -> load files to dictionary of DFs

Maybe we just need:

find files -> load files to dictionary of DFs -> add all other DFs that we are pointing to directly

Reply:
@brynpickering Thanks for the comment on this -- simplifed everything a bit, there wasn't really any need to restrict to either all csv or df. I've made these changes (see message below).